scripts/packages/VSCodeServer/src/trees.jl

```diff
@@ -80,28 +80,54 @@ end
 
 getfield_safe(x, f, default=UNDEF) = isdefined(x, f) ? getfield(x, f) : default
 
-function treerender(x)
+struct PropertyBox
+    object
+end
+treerender(x::PropertyBox) = invoke(treerender, Tuple{Any,String,String}, x.object, "#properties", wsicon(x))
+
+function treerender(x, typedisplay=typeof(x), icon=wsicon(x))
     fields = fieldnames(typeof(x))
 
     if isempty(fields)
-        treerender(Text(string(typeof(x), "()")))
+        treerender(Text(string(typedisplay, "()")))
     else
-        treerender(LazyTree(string(typeof(x)), wsicon(x), function ()
-            collect([SubTree(string(f), wsicon(getfield_safe(x, f)), getfield_safe(x, f)) for f in fields])
-        end))
+        treerender(LazyTree(string(typedisplay), icon,
+            () -> [SubTree(string(f), wsicon(getfield_safe(x, f)), getfield_safe(x, f)) for f in fields]
+        ))
     end
 end
 
-function treerender(x::Dict{K,V}) where {K,V}
-    treerender(LazyTree(string(nameof(typeof(x)), "{$(K), $(V)} with $(pluralize(length(keys(x)), "element", "elements"))"), wsicon(x), length(keys(x)) == 0, function ()
-        if length(keys(x)) > MAX_PARTITION_LENGTH
-            partition_by_keys(x, sz=MAX_PARTITION_LENGTH)
-        else
-            collect([SubTree(repr(k), wsicon(v), v) for (k, v) in x])
+function treerender(x::AbstractDict{K,V}) where {K,V}
+    treerender(LazyTree(string(nameof(typeof(x)), "{$(K), $(V)} with $(pluralize(length(keys(x)), "element", "elements"))"), wsicon(x), length(keys(x)) == 0,
+        () -> pushfirst!(
+            if length(keys(x)) > MAX_PARTITION_LENGTH
+                partition_by_keys(x, sz=MAX_PARTITION_LENGTH)
+            else
+                # collect is necessary because the return type of an array comprehension depends on the iterator,
+                # but we only want Arrays here
+                collect([SubTree(repr(k), wsicon(v), v) for (k, v) in x])
+            end,
+            SubTree("", wsicon(x), PropertyBox(x)),
+        )
+    ))
+end
+
+function treerender(x::AbstractArray{T,N}) where {T,N}
+    treerender(LazyTree(string(typeof(x), " with $(pluralize(size(x), "element", "elements"))"), wsicon(x), length(x) == 0,
+        function ()
+            out = if length(x) > MAX_PARTITION_LENGTH
+                partition_by_keys(x, sz=MAX_PARTITION_LENGTH)
+            else
+                # collect is necessary because the return type of an array comprehension depends on the iterator,
+                # but we only want Arrays here
+                collect([SubTree(repr(k), wsicon(v), v) for (k, v) in zip(keys(x), vec(assign_undefs(x)))])
+            end
+            x isa Array ? out : pushfirst!(out, SubTree("", wsicon(x), PropertyBox(x)))
```

Comment:
No need to show an empty `#properties` item for `Base.Array`s, so let's special case that here.